File: security-application/src/main/java/com/example/security/application/port/in/QueryAuditLogUseCase.java
package com.example.security.application.port.in;

import com.example.security.domain.audit.AuditEntry;
import com.example.security.domain.audit.AuditEntry.AuditAction;
import com.example.security.domain.common.TenantId;
import java.time.Instant;
import java.util.List;
import java.util.Optional;

/**
 * use case 8 — ISMS-P 2.9 통제. 누가 언제 어떤 검색 / 룰 변경 / 알람 처리 했는지 audit 조회.
 */
public interface QueryAuditLogUseCase {

  List<AuditEntry> query(AuditQuery query, OperatorContext operator);

  record AuditQuery(
      TenantId tenantId,
      Optional<String> actor,
      Optional<AuditAction> action,
      Optional<Instant> from,
      Optional<Instant> to,
      int size) {

    /** audit 조회 페이지 크기 상한 — 무제한 dump 방지 (API4 — 자원 소비 제한). */
    public static final int MAX_SIZE = 1000;

    public AuditQuery {
      if (size < 1 || size > MAX_SIZE) {
        throw new IllegalArgumentException("size 는 1~" + MAX_SIZE + ": " + size);
      }
    }
  }
}

File: security-application/src/main/kotlin/com/example/security/application/port/in/QueryAuditLogUseCase.kt
package com.example.security.application.port.`in`

import com.example.security.domain.audit.AuditEntry
import com.example.security.domain.common.TenantId
import java.time.Instant
import java.util.Optional

/**
 * use case 8 — ISMS-P 2.9 통제. 누가 언제 어떤 검색 / 룰 변경 / 알람 처리 했는지 audit 조회.
 */
interface QueryAuditLogUseCase {

    fun query(query: AuditQuery, operator: OperatorContext): List<AuditEntry>

    @JvmRecord
    data class AuditQuery(
        val tenantId: TenantId,
        val actor: Optional<String>,
        val action: Optional<AuditEntry.AuditAction>,
        val from: Optional<Instant>,
        val to: Optional<Instant>,
        val size: Int,
    ) {

        init {
            require(size in 1..MAX_SIZE) { "size 는 1~$MAX_SIZE: $size" }
        }

        companion object {
            /** audit 조회 페이지 크기 상한 — 무제한 dump 방지 (API4 — 자원 소비 제한). */
            const val MAX_SIZE: Int = 1000
        }
    }
}
